source/common/http/conn_manager_impl.cc

```diff
@@ -511,34 +511,37 @@ void ConnectionManagerImpl::ActiveStream::decodeHeaders(HeaderMapPtr&& headers,
 
   ASSERT(!cached_route_.valid());
   cached_route_.value(snapped_route_config_->route(*request_headers_, stream_id_));
-
-  // Check for WebSocket upgrade request if the route exists, and supports WebSockets.
-  // TODO if there are no filters when starting a filter iteration, the connection manager
-  // should return 404. The current returns no response if there is no router filter.
-  if ((protocol == Protocol::Http11) && cached_route_.value()) {
+  if (cached_route_.value()) {
     const Router::RouteEntry* route_entry = cached_route_.value()->routeEntry();
-    const bool websocket_allowed = (route_entry != nullptr) && route_entry->useWebSocket();
-    const bool websocket_requested = Utility::isWebSocketUpgradeRequest(*request_headers_);
-
-    if (websocket_requested && websocket_allowed) {
-      ENVOY_STREAM_LOG(debug, "found websocket connection. (end_stream={}):", *this, end_stream);
-
-      connection_manager_.ws_connection_.reset(new WebSocket::WsHandlerImpl(
-          *request_headers_, request_info_, *route_entry, *this,
-          connection_manager_.cluster_manager_, connection_manager_.read_callbacks_));
-      connection_manager_.ws_connection_->onNewConnection();
-      connection_manager_.stats_.named_.downstream_cx_websocket_active_.inc();
-      connection_manager_.stats_.named_.downstream_cx_http1_active_.dec();
-      connection_manager_.stats_.named_.downstream_cx_websocket_total_.inc();
-      return;
-    } else if (websocket_requested) {
-      // Do not allow WebSocket upgrades if the route does not support it.
-      connection_manager_.stats_.named_.downstream_rq_ws_on_non_ws_route_.inc();
-      HeaderMapImpl headers{{Headers::get().Status, std::to_string(enumToInt(Code::Forbidden))}};
-      encodeHeaders(nullptr, headers, true);
-      return;
+    request_info_.route_entry_ = route_entry;
```

Comment:
This should actually be done in the context of the route load call: https://github.com/envoyproxy/envoy/blob/master/source/common/http/conn_manager_impl.cc#L1150

This is because filters can change the route. So if there route ever changes it should be reset in the request info. Please also add a test for this case in conn_manager_impl_test

Reply:
I kept the assignment here and added the assignment in the `if` in `route` as you suggested.  I chose this way because:
1.  If we were to only do it in the load call, to ensure we get the value when no filter changes it, we'd need to set the value in RequestInfo on every `route()` rather than just the ones where we do a reload of the route (we'd have to do it outside the `if`).
2.  It seems strange to disconnect setting the `RequestInfoImpl` variable from setting the cache variable.  In the case where we only set this variable in `route()`, if no filter were to ever load the route (I don't know enough about the way requests are routed to know if this is possible), the `RequestInfo` would never be populated even though a route was technically resolved for the request.

Anyway, I wasn't exactly sure which way you were suggesting in the comment, but this way seemed more sensible.  If this isn't preferable, I'm happy to move it to only being set in `route()` as I'm not too familiar with the common/possible codepaths here.

Reply:
Sorry I didn't look at this carefully enough. Here is what I think I would do. I would add a new private function called `refreshCachedRoute()` and call that both inside `route() as well as in this function. Within `refreshCachedRoute()` you can set RequestInfo. Note also that you are check of route being not-nullptr in `route()` is not correct. If a route is changed which then results in nullptr, it should be reflected in RequestInfo. With that said, there is a sequence that you identify that looks like:

1) Set/cache initial route
2) Clear route cache
3) No one loads route() again forcing a new caching.

In this case I would argue it's fine to leave the RequestInfo route as it initially was.

Reply:
Sounds good.  Yeah, I was thinking of a private function, too.  Good point about setting `RouteEntry` to `nullptr` when `Route` is `nullptr`.  I'll get that cleaned up.